File: src/roar_net_api/algorithms/rls.py
# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0

from logging import getLogger
from time import perf_counter
from typing import Protocol, TypeVar

from ..operations import (
    SupportsApplyMove,
    SupportsLocalNeighbourhood,
    SupportsObjectiveValue,
    SupportsObjectiveValueIncrement,
    SupportsRandomMovesWithoutReplacement,
)
from ..utils.logging import PerformanceLogger

log = getLogger(__name__)

perflogger = PerformanceLogger()


class _Solution(SupportsObjectiveValue, Protocol): ...


_TSolution = TypeVar("_TSolution", bound=_Solution)


class _Move(SupportsApplyMove[_TSolution], SupportsObjectiveValueIncrement[_TSolution], Protocol): ...


class _Neighbourhood(SupportsRandomMovesWithoutReplacement[_TSolution, _Move[_TSolution]], Protocol): ...


class _Problem(SupportsLocalNeighbourhood[_Neighbourhood[_TSolution]], Protocol): ...


def rls(
    problem: _Problem[_TSolution],
    solution: _TSolution,
    budget: float,
) -> _TSolution:
    start = perf_counter()

    neigh = problem.local_neighbourhood()

    if perflogger.active:
        perflogger.log(solution.objective_value(), __name__)

    while perf_counter() - start < budget:
        for move in neigh.random_moves_without_replacement(solution):
            incr = move.objective_value_increment(solution)
            assert incr is not None
            if incr <= 0:
                log.info(f"Found increment: {incr}")
                solution = move.apply_move(solution)
                if perflogger.active:
                    perflogger.log(solution.objective_value(), __name__)
                break
            if perf_counter() - start >= budget:
                return solution
        else:
            break

    return solution

File: julia/src/algorithms/rls.jl

# SPDX-FileCopyrightText: © 2025 Authors of the roar-net-api-py project <https://github.com/roar-net/roar-net-api-py/blob/main/AUTHORS>
#
# SPDX-License-Identifier: Apache-2.0
"""
    rls(problem::Problem, solution::Solution, budget::Real) -> solution::Solution

Random Local Search: repeatedly sample random moves from the local
neighbourhood and apply any non-worsening move, until `budget` seconds
have elapsed or no improving move is found.
"""
function rls(problem::Problem, solution::Solution, budget::Real)::Solution
    start = time()

    neigh = local_neighbourhood(problem)

    while time() - start < budget
        found = false
        for mv in random_moves_without_replacement(neigh, solution)
            incr = objective_value_increment(mv, solution)
            @assert incr !== nothing
            if incr <= 0
                @info "Found increment: $incr"
                solution = apply_move(mv, solution)
                found = true
                break
            end
            if time() - start >= budget
                return solution
            end
        end
        if !found
            break
        end
    end

    return solution
end
